### PR TITLE
Case study for converting old ODE code to new ODE code (design-doc #19)

### PR DESCRIPTION
**Pull request by bbbales2**
@wds15 have a look at this.

I was working on converting the models [here](https://github.com/stan-dev/stat_comp_benchmarks) and thought we might as well write it somewhere. This is my attempt at that.

This feels rather barebones though.

Should there be more to this? Any suggestions?

### CODE REVIEW

**Comment by wds15** [comment]
will review this ... gimme a moment

File: knitr/convert-odes/convert_odes.Rmd [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+  }
+}
+```
+
+In the new interface the system function can be written:
+
+```{stan, output.var = "", eval = FALSE}
+functions {
+  vector rhs(real t, vector y, real alpha) {
+    vector[1] yp = -alpha * y;
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:59]: This is syntactically wrong. It should read `vector[1] yp = [-alpha * y]'; `

(or I would be surprised)

File: knitr/convert-odes/convert_odes.Rmd [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+    return yp;
+  }
+}
+```
+
+The new interface avoids unused arguments `x_r`, and `x_i`, and the
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:66]: "The new interface avoids any unused arguments such as `x_r`, and `x_i` in this example, and ..."

File: knitr/convert-odes/convert_odes.Rmd [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+
+`ode_bdf`, `ode_adams`, `ode_rk45` and `ode_bdf_tol`,`ode_adams_tol`,
+`ode_rk45_tol`
+
+The solvers in the first columns have default tolerance settings. The solvers
+in the second column accept arguments for relative tolerance, absolute
+tolerance, and the maximum number of steps to take between output times.
+
+This is different from the old interface where tolerances are presented
+as overloads to the same function.
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:86]: user may not be familiar with "overloads"... maybe say

"This is different from the old interface where tolerances are presented through using the same function name with a few more arguments."

File: knitr/convert-odes/convert_odes.Rmd [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+The arguments are:
+
+1. ```f``` - ODE system function
+
+2. ```y0``` - Initial state of the ODE
+
+3. ```t0``` - Initial time of the ODE
+
+4. ```times``` - Sorted array of times to which the ode will be solved (each
+  element must be greater than t0)
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:107]: maybe emphasise that we now allow successive times to be the same. So maybe say

"Sorted (and not strictly sorted) array..."

Comment (wds15) [reply knitr/convert-odes/convert_odes.Rmd:107]: Although... the old interface now allows that as well... whatever...

File: knitr/convert-odes/convert_odes.Rmd [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+1. ```t``` - Time at which to evaluate the ODE system function
+
+2. ```y``` - State at which to evaluate the ODE system function
+
+3. ```arg1, arg2, ...``` - Arguments passed unmodified from the ODE solver
+function call. The ```...``` must match the types of the corresponding
+```...``` arguments of the ODE solver function call.
+
+A call to `ode_bdf` returns the solution of the ODE specified by the right hand
+side (`f`) and the initial conditions (`y0` and `t0`) at the time points given
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:131]: "right hand side" => system function (stick with it, I would say)

File: knitr/convert-odes/convert_odes.Rmd [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+`ode_rk45_tol`/`ode_bdf_tol`/`ode_adams_tol` interfaces with defaults for
+`rel_tol`, `abs_tol`, and `max_num_steps`. For the RK45 solver the defaults
+are $10^{-6}$ for `rel_tol` and `abs_tol` and $10^6$ for `max_num_steps`.
+For the BDF/Adams solvers the defaults are $10^{-10}$ for `rel_tol` and
+`abs_tol` and $10^8$ for `max_num_steps`.
+
+For more detailed information about either interface, look at the function
+reference guide:
+[Old interface](https://mc-stan.org/docs/2_23/functions-reference/functions-ode-solver.html),
+[New interface](https://mc-stan.org/docs/2_23/functions-reference/functions-ode-solver.html)
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:178]: I think the links need an update (I don't know why this comment refers to many lines, I mean 177).

File: knitr/convert-odes/convert_odes.Rmd [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+## SIR Model
+
+### ODE System Function
+
+In the old SIR system function, `beta`, `kappa`, `gamma`, `xi`, and `delta`,
+are packed into the `real[] theta` argument. `kappa` isn't actually a model
+parameter so it is not clear why it is packed in with the other parameters,
+but it is. Promoting `kappa` to a parameter causes there to be more states
+in the extended ODE sensitivity system (used to get gradients of the ODE
+with respect to inputs). The ODE system function looks like:
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:194]: "(used to get gradients of the ODE
 with respect to inputs) and augmenting the ODE system function by unnecessary parameters must be strictly avoided for good performance."

This point is super duper important to say it very loud: Don't throw too many parameters into the system function .

File: knitr/convert-odes/convert_odes.Rmd [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+In the new ODE interface each of the arguments is appended on to the ODE
+solver function call. The RK45 ODE solver with default tolerances is called
+`ode_rk45`. Because the states are handled as `vector` variables, the solver
+output is an array of vectors (`vector[]`).
+
+```{stan, output.var = "", eval = FALSE}
+transformed parameters {
+  vector<lower=0>[4] y[N_t];
+  {
+    y = ode_rk45(simple_SIR, y0, t0, t, beta, kappa, gamma, xi, delta);
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:273]: we don't need the inner block any more. So just write

`vector<lower=0>[4] y[N_t] = ode_rk45(simple_SIR, y0, t0, t, beta, kappa, gamma, xi, delta);`

and maybe emphasise that no more temporary is needed for packing purposes.

File: knitr/convert-odes/convert_odes.Rmd [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+  }
+}
+```
+
+### Full Model
+
+The full model with the new interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/sir/sir.stan);
+the full model with the old interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/sir/sir.stan),
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:283]: the links point to the same files?

File: knitr/convert-odes/convert_odes.Rmd [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+
+In the new interface the arguments are simply passed at the end of the
+`ode_bdf` call:
+
+```
+transformed parameters {
+  vector[1] C[N_t];
+  {
+    real theta[3] = {k_a, K_m, V_m};
+    C = ode_bdf(one_comp_mm_elim_abs, C0, t0, times, k_a, K_m, V_m, D, V);
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:374]: let's get rid of the inner block and do

`vector[1] C[N_t] = ode_bdf(one_comp_mm_elim_abs, C0, t0, times, k_a, K_m, V_m, D, V);`

File: knitr/convert-odes/convert_odes.Rmd [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+  }
+}
+```
+### Full Model
+
+The full model with the new interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/pkpd/one_comp_mm_elim_abs.stan);
+the full model with the old interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/pkpd/one_comp_mm_elim_abs.stan),
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:383]: links correct?

File: knitr/convert-odes/convert_odes.Rmd [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+
+### Generated Quantities Only Model
+
+There is an additional PKPD model that only uses generated quantities. It has
+been converted to the new interface as well.
+
+The full model with the new interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan);
+the full model with the old interface can be found
+[here](https://github.com/stan-dev/stat_comp_benchmarks/blob/master/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan),
```
Comment (wds15) [inline knitr/convert-odes/convert_odes.Rmd:395]: links?

File: knitr/convert-odes/convert_odes.Rmd (hunk 1 above)
Comment (bbbales2) [reply knitr/convert-odes/convert_odes.Rmd:59]: y is a vector so scalar times a vector is still a vector (I checked this compiles)

File: knitr/convert-odes/convert_odes.Rmd (hunk 6 above)
Comment (bbbales2) [reply knitr/convert-odes/convert_odes.Rmd:178]: These links will be off until the 2.24 docs get built.

File: knitr/convert-odes/convert_odes.Rmd (hunk 9 above)
Comment (bbbales2) [reply knitr/convert-odes/convert_odes.Rmd:283]: Should point to the new model here now: https://github.com/stan-dev/stat_comp_benchmarks/pull/2

**Comment by bbbales2** [comment]
@wds15 this won't quite be finished until the 2_24 manual goes up and these changes to the test models go through: https://github.com/stan-dev/stat_comp_benchmarks/pull/2

**Comment by bbbales2** [comment]
@wds15 oh yeah Andrew wanted me to do authors so your name (and mine) is on this now FYI

**Comment by rok-cesnovar** [comment]
Can we merge this? It would be nice to have this online for the release candidate.

**Comment by wds15** [comment]
Yup. From my end we can merge.

**Comment by rok-cesnovar** [comment]
Not sure when this renders to the web page now.

**Pull request merged**